dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,276 @@
+            }
+            else
+            {
+                while (!operatorStack.empty() && operatorStack.top().Char != L'(')
+                {
+                    // Pop operators onto output until we reach a left paren
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+
+                if (operatorStack.empty())
+                {
+                    // Broken parenthesis
+                    return {};
+                }
+
+                // Pop left paren and discard
+                operatorStack.pop();
+            }
+        }
+    }
+
+    // Pop all remaining operators.
+    while (!operatorStack.empty())
+    {
+        if (operatorStack.top().Type == MathTokenType::Parenthesis)
+        {
+            // Broken parenthesis
+            return {};
+        }
+
+        postfixTokens.push_back(operatorStack.top());
+        operatorStack.pop();
+    }
+
+    return postfixTokens;
+}
+
+winrt::IReference<double> NumberBoxParser::ComputePostfixExpression(std::vector<MathToken> tokens)
+{
+    std::stack<double> stack;
+
+    for (auto const token : tokens)
+    {
+        if (token.Type == MathTokenType::Operator)
+        {
+            // There has to be at least two values on the stack to apply
+            if (stack.size() < 2)
+            {
+                return nullptr;
+            }
+
+            const auto op1 = stack.top();
+            stack.pop();
+
+            const auto op2 = stack.top();
+            stack.pop();
+
+            double result;
+
+            switch (token.Char)
+            {
+                case L'-':
+                    result = op2 - op1;
+                    break;
+
+                case L'+':
+                    result = op1 + op2;
+                    break;
+
+                case L'*':
+                    result = op1 * op2;
+                    break;
+
+                case L'/':
+                    if (op1 == 0)
+                    {
+                        // divide by zero
+                        return std::numeric_limits<double>::quiet_NaN();
+                    }
+                    else
+                    {
+                        result = op2 / op1;
+                    }
+                    break;
+
+                case L'^':
+                    result = std::pow(op2, op1);
+                    break;
+
+                default:
+                    return nullptr;
+            }
+
+            stack.push(result);
+        }
+        else if (token.Type == MathTokenType::Numeric)
+        {
+            stack.push(token.Value);
+        }
+    }
+
+    // If there is more than one number on the stack, we didn't have enough operations, which is also an error.
+    if (stack.size() != 1)
+    {
+        return nullptr;
+    }
+
+    return stack.top();
+}
+
+winrt::IReference<double> NumberBoxParser::Compute(const std::wstring_view expr, winrt::INumberParser numberParser)
```

Comment:
>winrt::IReference<double> [](start = 0, length = 25)

looks like the calling code doesn't need to pass this to any other WinRT components (like out as an answer through a property) so i would recommend using the "language" type, which is std::optional. It works just like IReference and is promotable to `IReference<double>` implicitly in the same way that std::wstring can be promoted to std::hstring when it needs to be. std::optional is just a struct so it's less expensive than IReference which is at least a heap allocation.

Reply:
It's like that because numberParser.ParseDouble returns an IReference<double>, so that way I can handle the result of both with the same code. If you have magic that will turn either one of them into the other please tell me because I couldn't do it. :(